File: base/src/vsdk/toolkit/environment/geometry/PolyhedralBoundedSolidValidationEngine.java
package vsdk.toolkit.environment.geometry;

import java.util.ArrayList;

import vsdk.toolkit.common.VSDK;

public class PolyhedralBoundedSolidValidationEngine
{
    private PolyhedralBoundedSolidValidationEngine()
    {
    }

    public static boolean validateIntermediate(PolyhedralBoundedSolid solid)
    {
        StringBuilder msg = new StringBuilder();
        boolean ok = true;
        PolyhedralBoundedSolidNumericPolicy.ToleranceContext numericContext =
            PolyhedralBoundedSolidNumericPolicy.forSolid(solid);

        ArrayList<PolyhedralBoundedSolidValidationStrategy> strategies =
            new ArrayList<PolyhedralBoundedSolidValidationStrategy>();
        strategies.add(new GeometricPlanarityStrategy());
        strategies.add(new TopologicalIntegrityStrategy());

        int i;
        for ( i = 0; i < strategies.size(); i++ ) {
            if ( !strategies.get(i).validate(solid, numericContext, msg) ) {
                ok = false;
                break;
            }
        }

        solid.setValidationState(ok);
        if ( !ok ) {
            VSDK.reportMessage(solid, VSDK.WARNING, "validateIntermediate",
                "Solid validation test failed!:\n" + msg.toString());
        }
        return ok;
    }

    public static boolean validateStrict(PolyhedralBoundedSolid solid)
    {
        StringBuilder msg = new StringBuilder();
        boolean ok = true;
        PolyhedralBoundedSolidNumericPolicy.ToleranceContext numericContext =
            PolyhedralBoundedSolidNumericPolicy.forSolid(solid);

        ArrayList<PolyhedralBoundedSolidValidationStrategy> strategies =
            new ArrayList<PolyhedralBoundedSolidValidationStrategy>();
        strategies.add(new GeometricPlanarityStrategy());
        strategies.add(new TopologicalIntegrityStrategy());
        strategies.add(new GeometricStrictLoopsStrategy());
        strategies.add(new GeometricStrictFaceIntersectionsStrategy());

        int i;
        for ( i = 0; i < strategies.size(); i++ ) {
            if ( !strategies.get(i).validate(solid, numericContext, msg) ) {
                ok = false;
                break;
            }
        }

        solid.setValidationState(ok);
        if ( !ok ) {
            VSDK.reportMessage(solid, VSDK.WARNING, "validateStrict",
                "Solid validation test failed!:\n" + msg.toString());
        }
        return ok;
    }

    private static class GeometricPlanarityStrategy
        implements PolyhedralBoundedSolidValidationStrategy
    {
        @Override
        public boolean validate(PolyhedralBoundedSolid solid,
            PolyhedralBoundedSolidNumericPolicy.ToleranceContext numericContext,
            StringBuilder msg)
        {
            return PolyhedralBoundedSolidGeometricValidator
                .validateAllFacesPlanarityAndPlanes(solid, numericContext, msg);
        }
    }

    private static class TopologicalIntegrityStrategy
        implements PolyhedralBoundedSolidValidationStrategy
    {
        @Override
        public boolean validate(PolyhedralBoundedSolid solid,
            PolyhedralBoundedSolidNumericPolicy.ToleranceContext numericContext,
            StringBuilder msg)
        {
            PolyhedralBoundedSolidTopologicalValidator
                .remakeEmanatingHalfedgesReferences(solid);
            if ( !PolyhedralBoundedSolidTopologicalValidator
                .validateTopologicalIntegrity(solid) ) {
                msg.append("  - Topological integrity test failed.\n");
                return false;
            }
            return true;
        }
    }

    private static class GeometricStrictLoopsStrategy
        implements PolyhedralBoundedSolidValidationStrategy
    {
        @Override
        public boolean validate(PolyhedralBoundedSolid solid,
            PolyhedralBoundedSolidNumericPolicy.ToleranceContext numericContext,
            StringBuilder msg)
        {
            return PolyhedralBoundedSolidGeometricValidator
                .validateLoopsStrict(solid, numericContext, msg);
        }
    }

    private static class GeometricStrictFaceIntersectionsStrategy
        implements PolyhedralBoundedSolidValidationStrategy
    {
        @Override
        public boolean validate(PolyhedralBoundedSolid solid,
            PolyhedralBoundedSolidNumericPolicy.ToleranceContext numericContext,
            StringBuilder msg)
        {
            return PolyhedralBoundedSolidGeometricValidator
                .validateFaceIntersectionsStrict(solid, numericContext, msg);
        }
    }
}

File: cpp/base/src/main/vsdk/toolkit/environment/geometry/volume/polyhedralBoundedSolid/PolyhedralBoundedSolidValidationEngine.cpp
#include "PolyhedralBoundedSolidValidationEngine.h"

#include "PolyhedralBoundedSolid.h"
#include "PolyhedralBoundedSolidGeometricValidator.h"
#include <cmath>

bool PolyhedralBoundedSolidValidationEngine::validateIntermediate(PolyhedralBoundedSolid* solid)
{
    if ( solid == 0 ) return false;
    bool ok = PolyhedralBoundedSolidGeometricValidator::validateUniqueFaceAndVertexIds(solid, 0);
    solid->setValidationState(ok);
    return ok;
}

bool PolyhedralBoundedSolidValidationEngine::validateBooleanInputs(PolyhedralBoundedSolid* solidA, PolyhedralBoundedSolid* solidB)
{
    return validateIntermediate(solidA) && validateIntermediate(solidB);
}

bool PolyhedralBoundedSolidValidationEngine::validateStrict(PolyhedralBoundedSolid* solid)
{
    return validateIntermediate(solid);
}

bool PolyhedralBoundedSolidValidationEngine::areGeometricallyIdentical(PolyhedralBoundedSolid* a, PolyhedralBoundedSolid* b)
{
    if ( a == 0 || b == 0 ) return false;
    double* ma = a->getMinMax();
    double* mb = b->getMinMax();
    bool same = true;
    for (int i = 0; i < 6; ++i) {
        if ( std::abs(ma[i] - mb[i]) > 1e-9 ) { same = false; break; }
    }
    delete[] ma;
    delete[] mb;
    return same;
}
